src/ynEIGEN/AssetRegistry.sol

```diff
@@ -0,0 +1,330 @@
+        if (!_assetData[asset].active) {
+            revert AssetNotActiveOrNonexistent(address(asset));
+        }
+
+        uint256 balanceInPool = asset.balanceOf(address(this));
+        if (balanceInPool != 0) {
+            revert AssetBalanceNonZeroInPool(balanceInPool);
+        }
+
+        uint256 strategyBalance = eigenStrategyManager.getStakedAssetBalance(asset);
+        if (strategyBalance != 0) {
+            revert AssetBalanceNonZeroInStrategyManager(strategyBalance);
+        }
+
+        // Remove asset from the assets array
+        uint256 assetIndex = findAssetIndex(asset);
+
+        // Move the last element into the place to delete
+        assets[assetIndex] = assets[assets.length - 1];
+        assets.pop();
+
+        // Remove asset from the mapping
+        delete _assetData[asset];
+
+        emit AssetDeleted(address(asset), assetIndex);
+    }
+
+    /**
+     * @notice Finds the index of an asset in the assets array.
+     * @param asset The asset to find.
+     * @return uint256 The index of the asset.
+     */
+    function findAssetIndex(IERC20 asset) internal view returns (uint256) {
+        for (uint256 i = 0; i < assets.length; i++) {
+            if (assets[i] == asset) {
+                return i;
+            }
+        }
+        revert AssetNotFound(address(asset));
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ASSET VALUE  -------------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /**
+     * @notice This function calculates the total assets of the contract
+     * @dev It iterates over all the assets in the contract, gets the latest price for each asset from the oracle, 
+     * multiplies it with the balance of the asset and adds it to the total
+     * @return total The total assets of the contract in the form of uint
+     */
+    function totalAssets() public view returns (uint256) {
+        uint256 total = 0;
+
+        uint256[] memory depositedBalances = getAllAssetBalances();
+
+        for (uint256 i = 0; i < assets.length; i++) {
+            uint256 balanceInUnitOfAccount = convertToUnitOfAccount(assets[i], depositedBalances[i]);
+            total += balanceInUnitOfAccount;
+        }
+        return total;
+    }
+
+    /**
+     * @notice Retrieves the total balances of all assets managed by the contract, both held directly and managed through strategies.
+     * @dev This function aggregates the balances of each asset held directly by the contract and in each LSDStakingNode, 
+     * including those managed by strategies associated with each asset.
+     * @return assetBalances An array of the total balances for each asset, indexed in the same order as the `assets` array.
+     */
+    function getAllAssetBalances()
```

Comment:
note: we should thoroughly test this function with various scenario tests to ensure we are getting the correct balances, for both ours and eigen state.